libbeat/processors/timeseries/timeseries.go

```diff
@@ -0,0 +1,140 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timeseries
+
+import (
+	"strings"
+
+	"github.com/elastic/beats/libbeat/processors"
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type timeseriesProcessor struct {
+	dimensions map[string]interface{}
+	prefixes   []string
+}
+
+// NewTimeSeriesProcessor returns a processor to add timeseries info to events
+// Events are processed to extract all their dimensions (keyword fields that
+// hold a dimension of the metrics) and compute a hash of all their values into
+// `timeseries.instance` field.
+func NewTimeSeriesProcessor(fields common.Fields) (processors.Processor, error) {
+	dimensions := map[string]bool{}
+	prefixes := map[string]bool{}
+	populateDimensions("", dimensions, prefixes, fields)
+
+	// remove false values and convert to map where a nil value means
+	// it's a dimension
+	dimensionsNilDict := map[string]interface{}{}
+	for k, v := range dimensions {
+		if v {
+			dimensionsNilDict[k] = nil
+		}
+	}
+
+	// convert the prefix map to a list
+	prefixList := []string{}
+	for k := range prefixes {
+		prefixList = append(prefixList, k)
+	}
+
+	return &timeseriesProcessor{dimensions: dimensionsNilDict, prefixes: prefixList}, nil
+}
+
+func (t *timeseriesProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	if event.TimeSeries {
+		instanceFields := common.MapStr{}
+
+		// map all dimensions & values
+		for k, v := range event.Fields.Flatten() {
+			if t.isDimension(k) {
+				instanceFields[k] = v
+			}
+		}
+
+		h, err := hashstructure.Hash(instanceFields, nil)
+		if err != nil {
+			// this should not happen, keep the event in any case
+			return event, err
+		}
+		event.Fields["timeseries"] = common.MapStr{
+			"instance": h,
+		}
+	}
+
+	return event, nil
+}
+
+func (t *timeseriesProcessor) isDimension(field string) bool {
+	if _, ok := t.dimensions[field]; ok {
+		return true
+	}
+
+	// field matches any of the prefixes
+	for _, prefix := range t.prefixes {
+		if strings.HasPrefix(field, prefix) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// put all dimension fields in the given map for quick access
+func populateDimensions(prefix string, dimensions map[string]bool, prefixes map[string]bool, fields common.Fields) {
+	for _, f := range fields {
+		name := f.Name
+		if prefix != "" {
+			name = prefix + "." + name
+		}
+
+		if len(f.Fields) > 0 {
+			populateDimensions(name, dimensions, prefixes, f.Fields)
+			continue
+		}
+
+		if f.Type == "object" {
+			// everything with this prefix could be a dimension
+			name += "."
+			if _, ok := prefixes[name]; !ok || f.Overwrite {
+				prefixes[name] = isDimension(f)
+			}
+		} else {
+			if _, ok := dimensions[name]; !ok || f.Overwrite {
+				dimensions[name] = isDimension(f)
+			}
```

Comment:
Should we only add dimensions to these maps if they are actually dimensions? This way we wouldn't need to filter and convert them later in `NewTimeSeriesProcessor()`.

Reply:
i changed it to be like this because i didn't find a way to solve the problem of fields with `Overwrite: true` without it. These changes assumes that fields can come in random order. If fields instead were guaranteed to come `Overwrite: false` first and then `Overwrite: true` _after_, we could avoid this. If they can show up in random order, i think we need this. Happy to discuss of course. An outstanding question is what happens if there are two fields with `Overwrite: true`, which one take precedence?

Reply:
`E_TOOMANYOVERRIDES` :P I'm ok with how it works now, had to add 8238b29 for prefixes